packages/create-yoga/utils/example.ts

```diff
@@ -0,0 +1,108 @@
+import { promisify } from "util";
+
+const pipeline = promisify(Stream.pipeline);
+
+export type RepoInfo = {
+  username: string;
+  name: string;
+  branch: string;
+  filePath: string;
+};
+
+export async function isUrlOk(url: string): Promise<boolean> {
+  const res = await got.head(url).catch((e) => e);
+  return res.statusCode === 200;
+}
+
+export async function getRepoInfo(
+  url: URL,
+  examplePath?: string
+): Promise<RepoInfo | undefined> {
+  const [, username, name, t, _branch, ...file] = url.pathname.split("/");
+  const filePath = examplePath
+    ? examplePath.replace(/^\//, "")
+    : file.join("/");
+
+  // Support repos whose entire purpose is to be a NextJS example, e.g.
+  // https://github.com/:username/:my-cool-nextjs-example-repo-name.
+  if (t === undefined) {
+    const infoResponse = await got(
+      `https://api.github.com/repos/${username}/${name}`
+    ).catch((e) => e);
+    if (infoResponse.statusCode !== 200) {
+      return;
+    }
+    const info = JSON.parse(infoResponse.body);
+    return { username, name, branch: info["default_branch"], filePath };
+  }
+
+  // If examplePath is available, the branch name takes the entire path
+  const branch = examplePath
+    ? `${_branch}/${file.join("/")}`.replace(new RegExp(`/${filePath}|/$`), "")
+    : _branch;
+
+  if (username && name && branch && t === "tree") {
+    return { username, name, branch, filePath };
+  }
+}
+
+export function hasRepo({
+  username,
+  name,
+  branch,
+  filePath,
+}: RepoInfo): Promise<boolean> {
+  const contentsUrl = `https://api.github.com/repos/${username}/${name}/contents`;
+  const packagePath = `${filePath ? `/${filePath}` : ""}/package.json`;
+
+  return isUrlOk(contentsUrl + packagePath + `?ref=${branch}`);
+}
+
+export function existsInRepo(nameOrUrl: string): Promise<boolean> {
+  try {
+    const url = new URL(nameOrUrl);
+    return isUrlOk(url.href);
+  } catch {
+    return isUrlOk(
+      `https://api.github.com/repos/dotansimha/graphql-yoga/contents/examples/${encodeURIComponent(
+        nameOrUrl
+      )}`
+    );
+  }
+}
+
+export function downloadAndExtractRepo(
+  root: string,
+  { username, name, branch, filePath }: RepoInfo
+): Promise<void> {
+  return pipeline(
+    got.stream(
+      `https://codeload.github.com/${username}/${name}/tar.gz/${branch}`
+    ),
+    tar.extract(
+      { cwd: root, strip: filePath ? filePath.split("/").length + 1 : 1 },
+      [`${name}-${branch.replace(/\//g, "-")}${filePath ? `/${filePath}` : ""}`]
+    )
+  );
+}
+
+export function downloadAndExtractExample(
+  root: string,
+  name: string
+): Promise<void> {
+  if (name === "__internal-testing-retry") {
+    throw new Error("This is an internal example for testing the CLI.");
+  }
+
+  return pipeline(
+    got.stream(
+      "https://codeload.github.com/dotansimha/graphql-yoga/tar.gz/master"
+    ),
+    tar.extract({ cwd: root, strip: 3 }, [
+      `graphql-yoga-master/examples/${name}`,
```

Comment:
I am not sure whether we want to pull the examples from master 🤔

Maybe we can when we release a new yoga version create a git tag (e.g. `stable-examples`) and it will fetch the examples from there.

Maybe we should even publish the examples to NPM and pull them from there?

I see tight now that it is downloading the full tar and then extracting the example out of it 🤔  That seems like too much is downloaded to me.